### ISSUE TITLE
When a module is included in two different chunks, tree shake it differently for each chunk

### ISSUE THREAD

**Issue by jedwards1211** [webpack/webpack#7782]
## Feature request

Here is the bug-ish issue I'm running into:

https://github.com/jedwards1211/webpack-code-splitting-issue/tree/two-async-chunks

### `index.js`
```js
import('./chunk1.js')
import('./chunk2.js')
```
### `chunk1.js`
```js
import { Toggle } from 'react-powerplug';
console.log(Toggle)
```
### `chunk2.js`
```js
import { Value } from 'react-powerplug';
console.log(Value)
```
### `webpack.config.js`
```js
// webpack.config.js
module.exports = {
  mode: 'production',
  externals: {
    'react': 'React'
  },
}
```
webpack it on the command line

**What is the expected behavior?**
`dist/1.js` contains `Toggle` but not `Value`.
`dist/2.js` contains `Value` but not `Toggle`.

**Current behavior**
`dist/1.js` contains both `Toggle` and `Value`.
`dist/2.js` contains both `Toggle` and `Value`.

It seems that the tree shaking is applied to `react-powerplug` main module itself, irrespective of the individual chunks that are importing from it.

**What is motivation or use case for adding/changing the behavior?**
Many packages nowadays use rollup to bundle their code into a single JS module and distribute that one file (or maybe one bundle for ES5, one bundle for ES2015).  As such, I want to be confident that Webpack will eliminate dead code effectively when several different chunks of my app are importing from a rollup bundle.

**How should this be implemented in your opinion?**
Hopefully, tree-shaking can be performed on `react-powerplug` for the import in `chunk1.js`, and then a second, different round of tree-shaking can be performed for the import in `chunk2.js`.  Right now it seems like Webpack is only performing one tree-shaking operation on `react-powerplug` based upon all imports from it from all chunks in my app.

**Are you willing to work on this yourself?**
I would need a lot of guidance, as this seems like this might require a significant refactor

**Comment by sokra** [comment]
This optimization is not in general valid. The side-effects of `react-powerplug` must be only evaluated once. Different tree-shaking for each chunk would create two different variants of `react-powerplug`, which both need to be evaluated. This could change the behavior.

Here is a simplified example which cause this problem:

``` js
// index.js
(async () => {
  await import('./chunk1.js');
  await import('./chunk2.js');
}());

// chunk1.js
import { A } from "module";
console.log(A());

// chunk2.js
import { B } from "module";
console.log(B());

// module.js
let c = 1;
export const A = () => `A = ${c++}`;
export const B = () => `B = ${c++}`;
```

This will log `A = 1, B = 2`.

Assuming we would create two variants which are tree-shaked individually.

``` js
// module.js in chunk1
let c = 1;
export const A = () => `A = ${c++}`;

// module.js in chunk1
let c = 1;
export const B = () => `B = ${c++}`;
```

You may notice that there are now two `c`s in different scopes.

This will log `A = 1, B = 1`. 😢 

Note that side-effects could be more complex and also in dependencies.

---

But there is a way to tree-shake this in the way you want.

You need to opt-in into `"sideEffects": false` in package.json and reexport symbols from different modules.

``` js
// react-powerplug/index.js
export { Toggle } from "./Toggle";
export { Value } from "./Value";

// react-powerplug/package.json
...
"sideEffects": false
...
```

**Comment by jedwards1211** [comment]
Wait so that would mean if it declares `sideEffects: false`, there should also be no problem with tree shaking a single roll-up bundle two different ways, right?

**Comment by jedwards1211** [comment]
(without necessarily double-loading the dependencies)

**Comment by jedwards1211** [comment]
Also note that the code for both Toggle and Value wound up duplicated in two chunks anyway, so it could have cause repeat side effects

**Comment by jedwards1211** [comment]
I don't have control over whether people release packages with multiple modules or a single roll-up bundle, so I'm pretty concerned about how I'll be able to tree shake the latter.

**Comment by TrySound** [comment]
@sokra Is there a way to tell webpack about side effects from the code? Can't it just analyse that there is not reassignments so there is not side effects? The only effect here is State class which is marked as pure so there shouldn't be a problem with it.
https://unpkg.com/react-powerplug@1.0.0-rc.1/dist/react-powerplug.esm.js

**Comment by ale-grosselle** [comment]
@TrySound Have u find a good workaround? 

I Have the same problem with my project that include some esm rollup project

**Comment by TrySound** [comment]
I use rollup for prod.

**Comment by posva** [comment]
Hey, I think having the tree shaking for statically imported files included in async chunks could make it very easy for people to split their bundles. As you said it does require having pure functions in the file. In flux architectures like vuex and probably redux and mobx, everything needed by the function is injected or available through _context_. As we were thinking of the next iteration of Vuex, this could really help to make the experience for end developers fully transparent while still properly splitting business logic across js files.

Usually in Vuex, actions/mutations (mutations and actions are merged together in future version of Vuex) look this:

```js
// module.js
export const A = state => state.a++
export const B = state => state.b++
```

the state variable is declared in a different file and injected into the function at runtime. If there was a way to flag this file as pure or without side effects (I don't really know what would be the wording), this would allow statically importing different functions of the same file in different Component (in the case of Vue)
Those components can, later on, be lazily loaded at a router level, which is something we encourage for **every route** and frameworks like Nuxt do by default. Right now, if you happen to:

```js
// Home.vue
import { A } from './module'
```
and
```js
// OtherPage.vue
import { B } from './module'
```

and lazy load both `Home.vue` and `OtherPage.vue` at the routing level, you end up with 2 js files but both contain the whole `module.js` file, ending up in unused code which grows as the module grows with the app.

Just to be clear, **this is not how currently Vuex works** (I don't know for the other state libraries), but this is how a hypothetical architecture would look like:

```
.
├── actions.js -> root level actions
├── getters.js -> root level getters
├── index.js -> store instance (contains state and imports of other modules)
└── someModule
    ├── actions.js -> actions relative to module someModule
    ├── getters.js -> getters relative to module someModule
    └── index.js -> exports a Module instance imported by store or lazily loaded
```

then, actions and getters, which is where most of the business logic ends up, can be statically (or dynamically) imported wherever they are used, also improving upon typing btw

I tried adding a `sideEffects` flag per @TheLarkInn recommendation:

```js
// webpack.config.js
  module: {
    rules: [
      {
        test: /store\/\w+\/(actions|getters)\.js$/,
        sideEffects: false,
      },
    ],
  },
```

but it yields the same result.

I have a boiled down repository to showcase the thing, I don't think it's necessary as the code provided is enough but just in case, here it is: https://github.com/posva/store-code-split-demo

**Comment by Bobris** [comment]
Here is example how this problem is solved in Bobril-build:
Input: https://github.com/bobril/bbcore/tree/master/Lib/Bundler/Js/Tests/Inputs/TwoLazyLibsSharingLib

Output: https://github.com/bobril/bbcore/tree/master/Lib/Bundler/Js/Tests/Expected/TwoLazyLibsSharingLib

Look inside files with "cb-" prefix (it means compressed and beautified by uglify).

Basically shared module is changed to be lazy imported and placed in separate chunk. Because code splitter always see complete program, it can modify imports of these chunks to first import this new shared chunk. Side effects are then applied only once and just before first lazy import.

**Comment by luixo** [comment]
@sokra 
Hello!
Is there any work going on this? I have an issue on splitting my (side-effect free) utils into two chunks with correct DCE.
I got a pretty minimal example:
```ts
// index.ts (entrypoint)

const moduleA = import(/* webpackChunkName: 'module-a' */'./module-a');
const moduleB = import(/* webpackChunkName: 'module-b' */'./module-b');

function index(): void {
    Promise.all([moduleA, moduleB])
        .then(([moduleA, moduleB]) => {
            moduleA.default();
            moduleB.default();
        });
}
```

```ts
// module-a.ts

import {foo} from './utils';

function run(): void {
    foo();
}

export default run;
```
```ts
// module-b.ts

import {bar} from './utils';

function run(): void {
    bar();
}

export default run;
```

```ts
// utils.ts

export function foo(): void {
    console.log('foo');
}

export function bar(): void {
    console.log('bar');
}

export function baz(): void {
    console.log('baz');
}
```
Both chunks ('module-a' and 'module-b') are built with `foo` and `bar`, `baz` is (correctly) not included.
Can I mark `utils.ts` as side-effect free somehow?

**Comment by posva** [comment]
here is some activity to prevent this from being autoclosed

**Comment by alexpavlovich** [comment]
ping

**Comment by wopian** [comment]
Bump to remove inactive label

**Comment by vankop** [comment]
Should work with `webpack@next`. Feel free to report new issue with reproducible repo.

**Comment by sokra** [comment]
This one is not fixed as it's related to multiple async imports instead of entrypoints.

**Comment by vankop** [comment]
Ah, sorry..

**Comment by hansottowirtz** [comment]
Are there any workarounds for this issue except splitting up side-effect free files into multiple files?

**Comment by b-e-r-t-o** [comment]
@sokra the workaround is not working for me as I'm getting the 'default' added still. To build on the example above my module looks like this:

```
export function foo(): void {
    console.log('foo');
}

export function bar(): void {
    console.log('bar');
}

export default Class Baz() {
    constructor() {
        console.log('baz');
    }
}
```

When I go look at my main chunk I see there is a reference to 'default' even though I am importing the separate modules as you've suggested. 

```
/***/ "./app/utils?foo":
/*!***********************************************************!*\
  !*** ./app/utils?foo ***!
  \***********************************************************/
/*! exports provided: foo, default */
```

**Comment by jedwards1211** [comment]
@sokra now that I think about your example, there would even be a problem with stateless modules if they have to refer to classes within themselves, class definitions would have to be moved to some shared chunk to prevent `instanceof` problems.

**Comment by jedwards1211** [comment]
Actually I'm confused.  Looking back at my OP it seems like `Toggle` and `Value` instances already *do* get duplicated.  But I assume the first chunk to get loaded registers that module with webpack, and the second chunk loads that preregistered module instead of registering duplicate class, function, etc instances?

**Comment by sokra** [comment]
Yes exactly. At runtime there is a module cache that makes sure that only a single instance of each module is instantiated. The first loaded chunk will add the module to the cache and the second chunk will use the module from the cache. So the modules in the chunks need to be identical as any of them might be used for both.

**Comment by jedwards1211** [comment]
Makes sense.  Assuming we had code to split a stateless module into a base of common local variables used by all importing chunks, and tree-shaked modules that import from the base and export only what's used by the chunks, how tricky would it be to work those split modules into webpack?  I've dealt with fairly complex codemods and scope traversal with babel a lot, but I'm not very familiar with webpack internals.

**Comment by b-e-r-t-o** [comment]
@sokra will this come to Webpack? It looks like esbuild supports it now. Maybe that can be used as a reference point? https://github.com/evanw/esbuild/blob/d926cebcc3fffce6e12aa345ae741211353fe6e2/docs/architecture.md#code-splitting

**Comment by vankop** [comment]
Should work with `webpack@latest`. Feel free to report new issue with reproducible repo.

**Comment by dhlolo** [comment]
I saw that 

> Should work with `webpack@latest`. Feel free to report new issue with reproducible repo.
> 
> take a look into `runtime` feature

Do you mean optimization.usedExports &&  optimization.runtimeChunk ? If runtimeChunk is 'single', it seems that the 'side-effect safe' exports will not be divided into multi chunk.

**Comment by elado** [comment]
Is this supposed to work for a mix of regular and dynamic imports?

While I saw it working in https://github.com/vankop/webpack-runtime for multiple **entries**, when I have a single entry and some dynamic imports the file with the functions doesn't get split.

Example:

```js
// utils.js
export function square(x) { return x * x; }
export function cube(x) { return x * x * x; }


// index.js
import { cube } from './utils'

console.log(cube(4))

async function main() {
  import('./async')
}

main()

// async.js
import { square } from './utils'

console.log(square(4))
```

even when splitting utils.js I still see a single chunk of it.

```js
    splitChunks: {
      cacheGroups: {
        utils: {
          test: /utils/,
          usedExports: true,
          enforce: true,
          reuseExistingChunk: false,
        },
      },
    },
```

The result is another chunk for utils with both functions, which means that it has to be loaded on the initial load of the app.

**Comment by Taoja** [comment]
i meet this problem in my single page application and i written a plugin to resolution it: 
[split-module-webpack-plugin](https://github.com/Taoja/split-module-webpack-plugin)

@elado @jedwards1211 @sokra
